File: app/src/main/java/com/tuhoang/pocketmind/ui/home/HomeViewModel.java
package com.tuhoang.pocketmind.ui.home;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.tuhoang.pocketmind.data.models.Transaction;
import com.tuhoang.pocketmind.utils.AppLogger;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class HomeViewModel extends ViewModel {

    private final MutableLiveData<List<Transaction>> recentTransactions = new MutableLiveData<>();
    private final MutableLiveData<Double> totalMonthlyExpense = new MutableLiveData<>();
    private final MutableLiveData<Map<String, Double>> categoryExpenses = new MutableLiveData<>();

    private final FirebaseAuth mAuth = FirebaseAuth.getInstance();
    private final FirebaseFirestore db = FirebaseFirestore.getInstance();

    public LiveData<List<Transaction>> getRecentTransactions() {
        return recentTransactions;
    }

    public LiveData<Double> getTotalMonthlyExpense() {
        return totalMonthlyExpense;
    }

    public LiveData<Map<String, Double>> getCategoryExpenses() {
        return categoryExpenses;
    }

    public void fetchHomeData() {
        FirebaseUser user = mAuth.getCurrentUser();
        if (user == null) {
            totalMonthlyExpense.setValue(0.0);
            return;
        }

        String uid = user.getUid();

        // 1. Fetch Recent Transactions
        db.collection("users").document(uid).collection("expenses")
                .orderBy("timestamp", Query.Direction.DESCENDING)
                .limit(10)
                .get()
                .addOnSuccessListener(queryDocumentSnapshots -> {
                    List<Transaction> list = new ArrayList<>();
                    for (DocumentSnapshot doc : queryDocumentSnapshots) {
                        Transaction t = doc.toObject(Transaction.class);
                        if (t != null) list.add(t);
                    }
                    recentTransactions.setValue(list);
                })
                .addOnFailureListener(e -> AppLogger.e("HomeViewModel", "Failed to fetch recent transactions", e));

        // 2. Fetch Monthly Data
        Calendar cal = Calendar.getInstance();
        cal.set(Calendar.DAY_OF_MONTH, 1);
        cal.set(Calendar.HOUR_OF_DAY, 0);
        cal.set(Calendar.MINUTE, 0);
        cal.set(Calendar.SECOND, 0);
        Date startOfMonth = cal.getTime();

        db.collection("users").document(uid).collection("expenses")
                .whereGreaterThanOrEqualTo("timestamp", startOfMonth)
                .get()
                .addOnSuccessListener(queryDocumentSnapshots -> {
                    double totalExpense = 0;
                    Map<String, Double> categoryMap = new HashMap<>();

                    for (DocumentSnapshot doc : queryDocumentSnapshots) {
                        Transaction t = doc.toObject(Transaction.class);
                        if (t != null && "expense".equalsIgnoreCase(t.getType())) {
                            totalExpense += t.getAmount();
                            String cat = t.getCategory() != null ? t.getCategory() : "Other";
                            categoryMap.put(cat, categoryMap.getOrDefault(cat, 0.0) + t.getAmount());
                        }
                    }

                    totalMonthlyExpense.setValue(totalExpense);
                    categoryExpenses.setValue(categoryMap);
                })
                .addOnFailureListener(e -> AppLogger.e("HomeViewModel", "Failed to fetch monthly expenses", e));
    }
}

File: app/src/main/kotlin/com/tuhoang/pocketmind/ui/home/HomeViewModel.kt
package com.tuhoang.pocketmind.ui.home

import androidx.lifecycle.ViewModel
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.tuhoang.pocketmind.data.models.Transaction
import com.tuhoang.pocketmind.utils.AppLogger
import com.tuhoang.pocketmind.utils.PrefsManager
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.SharingStarted
import java.util.Calendar
import java.util.Date

enum class BudgetAlert { NONE, WARNING, EXCEEDED }

class HomeViewModel : ViewModel() {

    private val auth = FirebaseAuth.getInstance()
    private val db = FirebaseFirestore.getInstance()
    private val prefs = PrefsManager.getInstance()

    private val _allTransactions = MutableStateFlow<List<Transaction>>(emptyList())
    private val _totalMonthlyExpense = MutableStateFlow(0.0)
    val totalMonthlyExpense: StateFlow<Double> = _totalMonthlyExpense.asStateFlow()

    private val _categoryExpenses = MutableStateFlow<Map<String, Double>>(emptyMap())
    val categoryExpenses: StateFlow<Map<String, Double>> = _categoryExpenses.asStateFlow()

    private val _isLoading = MutableStateFlow(true)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    private val _categoryFilter = MutableStateFlow<String?>(null)
    val categoryFilter: StateFlow<String?> = _categoryFilter.asStateFlow()

    val monthlyBudget: StateFlow<Double> get() = budget

    private val _budgetValue = MutableStateFlow(prefs.getMonthlyBudget())
    val budget: StateFlow<Double> = _budgetValue.asStateFlow()

    val budgetAlert: StateFlow<BudgetAlert> = combine(_totalMonthlyExpense, _budgetValue) { spent, budget ->
        when {
            budget <= 0 -> BudgetAlert.NONE
            spent >= budget -> BudgetAlert.EXCEEDED
            spent >= budget * 0.8 -> BudgetAlert.WARNING
            else -> BudgetAlert.NONE
        }
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), BudgetAlert.NONE)

    val recentTransactions: StateFlow<List<Transaction>> = combine(
        _allTransactions,
        _searchQuery,
        _categoryFilter
    ) { list, query, category ->
        list.filter { tx ->
            val matchesQuery = query.isBlank() ||
                tx.category?.contains(query, ignoreCase = true) == true ||
                tx.note?.contains(query, ignoreCase = true) == true
            val matchesCategory = category.isNullOrBlank() || tx.category.equals(category, ignoreCase = true)
            matchesQuery && matchesCategory
        }
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val availableCategories: StateFlow<List<String>> = _allTransactions
        .map { list -> list.mapNotNull { it.category }.distinct().sorted() }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    fun setSearchQuery(query: String) { _searchQuery.value = query }
    fun setCategoryFilter(category: String?) { _categoryFilter.value = category }
    fun refreshBudget() { _budgetValue.value = prefs.getMonthlyBudget() }

    fun fetchHomeData(refreshing: Boolean = false) {
        val user = auth.currentUser ?: run {
            _isLoading.value = false
            _isRefreshing.value = false
            _totalMonthlyExpense.value = 0.0
            _allTransactions.value = emptyList()
            _categoryExpenses.value = emptyMap()
            return
        }

        if (refreshing) _isRefreshing.value = true else _isLoading.value = true
        _budgetValue.value = prefs.getMonthlyBudget()
        val uid = user.uid

        db.collection("users").document(uid).collection("expenses")
            .orderBy("timestamp", Query.Direction.DESCENDING)
            .limit(50)
            .get()
            .addOnSuccessListener { snapshots ->
                _allTransactions.value = snapshots.mapNotNull { it.toObject(Transaction::class.java) }
                finishLoading(refreshing)
            }
            .addOnFailureListener { e ->
                AppLogger.e("HomeViewModel", "Failed to fetch transactions", e)
                finishLoading(refreshing)
            }

        val cal = Calendar.getInstance().apply {
            set(Calendar.DAY_OF_MONTH, 1)
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
        }
        val startOfMonth: Date = cal.time

        db.collection("users").document(uid).collection("expenses")
            .whereGreaterThanOrEqualTo("timestamp", startOfMonth)
            .get()
            .addOnSuccessListener { snapshots ->
                var totalExpense = 0.0
                val categoryMap = mutableMapOf<String, Double>()

                for (doc in snapshots) {
                    val t = doc.toObject(Transaction::class.java) ?: continue
                    if (t.type.equals("expense", ignoreCase = true)) {
                        totalExpense += t.amount
                        val cat = t.category ?: "Other"
                        categoryMap[cat] = (categoryMap[cat] ?: 0.0) + t.amount
                    }
                }

                _totalMonthlyExpense.value = totalExpense
                _categoryExpenses.value = categoryMap
            }
            .addOnFailureListener { e ->
                AppLogger.e("HomeViewModel", "Failed to fetch monthly expenses", e)
            }
    }

    private fun finishLoading(refreshing: Boolean) {
        if (refreshing) _isRefreshing.value = false else _isLoading.value = false
    }
}
